backbone/src/test/java/org/researchstack/backbone/ui/task/TaskViewModelTest.kt

```diff
@@ -0,0 +1,419 @@
+        // Assemble
+        taskViewModel = createViewModel()
+
+        val currentStepMocked = createCurrentStep()
+        val previousStepMocked = createCurrentStep("previousStepMocked")
+        taskViewModel.currentStep = currentStepMocked
+        `when`(task!!.getStepBeforeStep(any(), any())).thenReturn(previousStepMocked)
+
+        // Act
+        taskViewModel.previousStep()
+
+        // Assert
+        Assert.assertEquals(taskViewModel.currentStep, previousStepMocked)
+        Assert.assertEquals(taskViewModel.currentStepEvent.value!!.step, previousStepMocked)
+    }
+
+    @Test
+    fun checkPrevious_shouldCloseWhenPreviousIsNull() {
+        // Assemble
+        taskViewModel = createViewModel()
+        `when`(task!!.getStepBeforeStep(any(), any())).thenReturn(null)
+        val currentStepMocked = createCurrentStep()
+        taskViewModel.currentStep = currentStepMocked
+
+        // Act
+        taskViewModel.previousStep()
+
+        // Assert
+        Assert.assertEquals(taskViewModel.taskCompleted.value, false)
+    }
+
+    @Test
+    fun testNext_shouldCloseWhenNextNull() {
+        // Assemble
+        taskViewModel = spy(createViewModel())
+
+        // Act
+        taskViewModel.nextStep()
+
+        // Assert
+        Assert.assertEquals(taskViewModel.taskCompleted.value, true)
+    }
+
+    @Test
+    fun testShowCurrentStep_ifCurrentStepIsNull_ShowGoToNextStep() {
+        // Assemble
+        taskViewModel = spy(createViewModel())
+        // Act
+        taskViewModel.showCurrentStep()
+
+        // Assert
+        verify(taskViewModel, times(1)).showCurrentStep()
+    }
+
+    @Test
+    fun testPreviousOnEditMode_shouldCheckForCancelDialog() {
+        // Assemble
+        taskViewModel = spy(createViewModel())
+        val currentStepMocked = createCurrentStep()
+        taskViewModel.edit(currentStepMocked)
+
+        // Act
+        taskViewModel.previousStep()
+
+        // Assert
+        verify(taskViewModel, times(1)).showCancelEditAlert()
+    }
+
+    @Test
+    fun testNextStepEditMode_WhenNextStepIsReviewStep_EditShouldBeFalse_moveReviewStepShouldGoToReviewStep() {
+        // Assemble
+        taskViewModel = spy(createViewModel())
+
+        val currentStepMocked = createCurrentStep()
+        val nextStepMocked = createCurrentStep("nextStepMocked")
+        taskViewModel.currentStep = currentStepMocked
+
+        `when`(task!!.getStepAfterStep(any(), any())).thenReturn(nextStepMocked)
+        `when`(taskViewModel.isReviewStep(nextStepMocked)).thenReturn(true)
+        taskViewModel.edit(nextStepMocked)
+
+        // Act
+        taskViewModel.nextStep()
+
+        // Assert
+        Assert.assertEquals(taskViewModel.editing, false)
+        Assert.assertEquals(taskViewModel.moveReviewStep.value?.step, nextStepMocked)
+    }
+
+    @Test
+    fun testCheckIfAnswersAreNotTheSame_false() {
+        // Assemble
+        taskViewModel = spy(createViewModel())
```

Comment:
I ask again, because i see it's a pattern here, why spy on the viewModel? Why not provide a valid set of steps and provide that to the viewModel instead? It may seem like more work, but then it would be more realistic; is it too complicated to initialize the view model with real data?